test/server_test.py

```diff
@@ -106,12 +112,41 @@ def missingCSRFToken():
 			})
 			assertResponse(response, 400, 'Session expired. Reload and try again')
 
+		@it('Empty body')
+		def emptyBody():
+			res1 = app.post('/login/form')
+			res2 = app.post('/login/form', data={})
+			res3 = app.post('/login/form', json={})
+			assertResponse(res1, 400, 'Session expired. Reload and try again')
+			assertResponse(res2, 400, 'Session expired. Reload and try again')
+			assertResponse(res3, 400, 'Session expired. Reload and try again')
+
+		@it('Non-json body')
+		def incorrectData():
+			res = app.post('/login/form', data='I am not json')
+			assertResponse(res, 400, 'Session expired. Reload and try again')
+
+		@it('Hitting rate limit')
+		def rateLimit():
+			enableRateLimiter(True)
+			data = {
+				'csrf_token': getLoginCSRFToken(),
+				'username': test_name,
+				'password': test_pass
+			}
+			for i in range(0, 10):
+				res = app.post('/login/form', data=data)
+				assertResponse(res, 200, 'Ok')
```

Comment:
I think it may be useful to make this assert print the number of tries it got to before it failed, if that ends up happening.

This would require extending assertResponse though. Hamcrest asserts allow for you to give a description message for their failure state. As seen here: [https://github.com/hamcrest/PyHamcrest/blob/master/examples/ExampleWithAssertThat.py](url)

You'd add a string argument like `"Prematurely failed at %d tries." % (i+1)`

Even if it ends up not being used much outside of cases like this, I still think it might be a good idea to keep that consistent.